src/error.rs

```diff
@@ -0,0 +1,391 @@
+//! Types related to error reporting.
+//!
+//! ## Single failure mode errors
+//!
+//! Generally speaking, zerocopy's conversions may fail for one of up to three reasons:
+//! - [`AlignmentError`]: the conversion source was improperly aligned
+//! - [`SizeError`]: the conversion source was of incorrect size
+//! - [`ValidityError`]: the conversion source contained invalid data
+//!
+//! Methods that only have one failure mode, like [`Ref::new_unaligned`], return
+//! that mode's corresponding error type directly.
+//!
+//! ## Compound errors
+//!
+//! Conversion methods that have either two or three possible failure modes
+//! return one of these error types:
+//! - [`CastError`]: the error type of ref-conversions.
+//! - [`TryCastError`]: the error type of fallible ref-conversions.
+//! - [`TryReadError`]: the error type of fallible read-conversions.
+//!
+//! ## Accessing the conversion source
+//!
+//! All error types provide an `into_src` method that converts the error into
+//! the source value underlying the failed conversion.
+
+use core::{convert::Infallible, fmt, marker::PhantomData, ops::Deref};
+
+#[cfg(doc)]
+use crate::{FromBytes, Ref, TryFromBytes};
+
+mod private {
+    #[cfg(doc)]
+    use super::*;
+
+    /// Zerocopy's generic error type.
+    ///
+    /// Generally speaking, zerocopy's conversions may fail for one of up to three reasons:
+    /// - [`AlignmentError`]: the conversion source was improperly aligned
+    /// - [`SizeError`]: the conversion source was of incorrect size
+    /// - [`ValidityError`]: the conversion source contained invalid data
+    ///
+    /// However, not all conversions produce all errors. For instance,
+    /// [`FromBytes::ref_from`] may fail due to alignment or size issues, but not
+    /// validity issues. This generic error type captures these (im)possibilities
+    /// via parameterization: `A` is parameterized with [`AlignmentError`], `S` is
+    /// parameterized with [`SizeError`], and `V` is parameterized with
+    /// [`Infallible`].
+    ///
+    /// Zerocopy never uses this type directly in its API. Rather, we provide three
+    /// pre-parameterized aliases:
+    /// - [`CastError`]: the error type of ref-conversions.
+    /// - [`TryCastError`]: the error type of fallible ref-conversions.
+    /// - [`TryReadError`]: the error type of fallible read-conversions.
+    #[derive(PartialEq, Eq)]
+    pub enum ConvertError<A, S, V> {
+        /// The conversion source was improperly aligned.
+        #[doc(hidden)]
+        Alignment(A),
+        /// The conversion source was of incorrect size.
+        #[doc(hidden)]
+        Size(S),
+        /// The conversion source contained invalid data.
+        #[doc(hidden)]
+        Validity(V),
+    }
+}
```

Comment:
@joshlf I've gone ahead and encapsulated the `ConvertError` in a private struct, so the public API consists solely of:
-  `AlignmentError<Src, Dst>`: the conversion source was improperly aligned
- `SizeError<Src, Dst>`: the conversion source was of incorrect size
- `ValidityError<Src, Dst>`: the conversion source contained invalid data
- `CastError<Src, Dst>`: the compound error type of ref-conversions
- `TryCastError<Src, Dst>`: the compound error type of fallible ref-conversions
- `TryReadError<Src, Dst>`: the compound error type of fallible read-conversions